contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,460 @@
+
+The existing Kubernetes volume components (attach/detach controller, PVC/PV controller, Kubelet volume manager) will handle the lifecycle of the CSI volume plugin operations (everything from triggering volume provisioning/deleting, attaching/detaching, and mounting/unmounting) just as they do for existing in-tree volume plugins.
+
+#### Proposed API
+
+A new `CSIPersistentVolumeSource` object will be added to the Kubernetes API. It will be part of the existing `PersistentVolumeSource` objects and thus can be used only via PersistentVolumes. For now we do not consider allowing CSI volumes directly from Pods without PersistentVolumeClaim.
+
+```GO
+type CSIPersistentVolumeSource struct {
+  // Driver is the name of the driver to use for this volume.
+  // Required.
+  Driver string `json:"driver" protobuf:"bytes,1,opt,name=driver"`
+
+  // VolumeHandle is the unique volume name returned by the CSI volume
+  // plugin’s CreateVolume to refer to the volume on all subsequent calls.
+  VolumeHandle string `json:"volumeHandle" protobuf:"bytes,2,opt,name=volumeHandle"`
+
+  // Optional: The value to pass to ControllerPublishVolumeRequest.
+  // Defaults to false (read/write).
+  // +optional
+  ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,5,opt,name=readOnly"`
+}
+```
+
+#### Internal Interfaces
+
+The in-tree CSI volume plugin will implement the following internal Kubernetes volume interfaces:
+
+1. `VolumePlugin`
+    * Mounting/Unmounting of a volume to a specific path.
+2. `AttachableVolumePlugin`
+    * Attach/detach of a volume to a given node.
+
+Notably, `ProvisionableVolumePlugin` and `DeletableVolumePlugin` are not implemented because provisioning and deleting for CSI volumes is handled by an external provisioner.
+
+#### Mount and Unmount
+
+The in-tree volume plugin’s SetUp and TearDown methods will trigger the `NodePublishVolume` and `NodeUnpublishVolume` CSI calls via Unix Domain Socket. Kubernetes will generate a unique `target_path` (unique per pod per volume) to pass via `NodePublishVolume` for the CSI plugin to mount the volume. Upon successful completion of the `NodeUnpublishVolume` call (once volume unmount has been verified), Kubernetes will delete the directory.
+
+The Kubernetes volume sub-system does not currently support block volumes (only file), so for alpha, the Kubernetes CSI volume plugin will only support file.
+
+#### Attaching and Detaching
+
+The attach/detach controller,running as part of the kube-controller-manager binary on the master, decides when a CSI volume must be attached or detached from a particular node.
+
+When the controller decides to attach a CSI volume, it will call the in-tree CSI volume plugin’s attach method. The in-tree CSI volume plugin’s attach method will do the following:
+
+1. Create a new `VolumeAttachment` object (defined in the “Communication Channels” section) to attach the volume.
+    * The name of the of the `VolumeAttachment` object will be the `pv-<PVName>-<NodeName>`.
```

Comment:
Have to handle name format overflows.   PV name can be a subdomain.  Node name can be a subdomain.  That can overflow a subdomain.  Unless you make the name format for an attachment be more than 2x a subdomain.  Not sure what the right answer is, but you have to spec it.

Reply:
No requirement that all names are subdomains (some are not) but the general default is subdomain or even label.  What does the name being defined this way buy us?  Can the name be made simpler, and the data be carried in the spec?